sympy/matrices/expressions/matexpr.py

```diff
@@ -821,6 +778,132 @@ def __nonzero__(self):
 def matrix_symbols(expr):
     return [sym for sym in expr.free_symbols if sym.is_Matrix]
 
+
+class _LeftRightArgs(object):
```

Comment:
If there is any reference for the algorithm can you add it somewhere as a comment or docstring.